File: legacy-rust/src/usage/orchestrator.rs
//! Orchestrates live usage fetches: walks atop's credentials for each supported
//! provider and returns the union of {@link UsageReport}s plus a per-credential
//! error log so the TUI can show *why* a provider is empty instead of looking
//! blank.
use anyhow::Result;
use std::collections::HashMap;
use std::sync::Arc;
use tokio::task::JoinSet;

use super::{fetcher_for, UsageReport};
use crate::auth::store as cred_store;
use crate::auth::Credential;
use crate::db::Db;
#[derive(Debug, Clone)]
pub struct FetchError {
    pub provider: String,
    pub account: String,
    pub message: String,
}

pub struct FetchResult {
    pub reports: Vec<UsageReport>,
    pub errors: Vec<FetchError>,
}

impl std::fmt::Display for FetchError {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(f, "{} ({}): {}", self.provider, self.account, self.message)
    }
}
pub struct Orchestrator {
    client: reqwest::Client,
}

impl Orchestrator {
    pub fn new() -> Self {
        Self { client: super::http::client() }
    }
    /// Fetch live usage for every credential atop has stored for `providers`.
    /// Per-credential errors are collected and returned (not silently dropped):
    /// a single failed provider must not blank the dashboard, but the user
    /// *must* see why a provider returned nothing.
    pub async fn fetch_all(&self, db: &Db, providers: &[&str]) -> Result<FetchResult> {
        let client = Arc::new(self.client.clone());
        let mut set: JoinSet<FetchOutcome> = JoinSet::new();
        let mut errors = Vec::new();
        for &pid in providers {
            let creds = cred_store::load(db, pid).unwrap_or_default();
            for cred in creds {
                // Refresh synchronously on the main task so JoinSet children
                // can own their credentials and never touch the Db. Refresh
                // is rare (only on token expiry) so the lack of fan-out here
                // is fine.
                let account = cred.account_label();
                let cred = match refresh(db, pid, &cred, &self.client).await {
                    Ok(Some(c)) => c,
                    Ok(None) => cred,
                    Err(e) => {
                        errors.push(FetchError {
                            provider: pid.to_string(),
                            account,
                            message: e.to_string(),
                        });
                        continue;
                    }
                };
                let pid = pid.to_string();
                let account = cred.account_label();
                let client = client.clone();
                let db_path = db.path.clone();
                set.spawn(async move {
                    let Some(fetcher) = fetcher_for(&pid) else {
                        return FetchOutcome::Err(FetchError {
                            provider: pid,
                            account,
                            message: "no fetcher registered for this provider".into(),
                        });
                    };
                    let db = crate::db::Db::open(&db_path).map_err(|e| FetchError {
                        provider: pid.clone(),
                        account: account.clone(),
                        message: e.to_string(),
                    });
                    let Ok(local_db) = db else { return FetchOutcome::Err(db.err().unwrap()) };
                    match fetcher.fetch_with_db(&local_db, &cred, &client).await {
                        Ok(Some(r)) => FetchOutcome::Ok(r),
                        Ok(None) => FetchOutcome::Err(FetchError {
                            provider: pid,
                            account,
                            message: "provider returned no usage data".into(),
                        }),
                        Err(e) => FetchOutcome::Err(FetchError {
                            provider: pid,
                            account,
                            message: e.to_string(),
                        }),
                    }
                });
            }
        }
        let mut reports = Vec::new();
        while let Some(res) = set.join_next().await {
            match res {
                Ok(FetchOutcome::Ok(r)) => reports.push(r),
                Ok(FetchOutcome::Err(e)) => errors.push(e),
                Err(join) => errors.push(FetchError {
                    provider: "?".into(),
                    account: "?".into(),
                    message: format!("task panicked: {join}"),
                }),
            }
        }
        Ok(FetchResult { reports, errors })
    }

    /// Per-provider, freshest report; used by the TUI.
    pub fn freshest(reports: &[UsageReport]) -> HashMap<String, UsageReport> {
        let mut by: HashMap<String, UsageReport> = HashMap::new();
        for r in reports {
            match by.get(&r.provider) {
                Some(prev) if prev.fetched_at >= r.fetched_at => {}
                _ => {
                    by.insert(r.provider.clone(), r.clone());
                }
            }
        }
        by
    }
}

enum FetchOutcome {
    Ok(UsageReport),
    Err(FetchError),
}

/// Refresh `cred` in place if it's an expired OAuth token with a refresh
/// token. Returns the new credential on success, or `None` if no refresh
/// was needed / possible. Pre-fetches on the calling task so the spawn
/// pool never touches the Db.
async fn refresh(
    db: &Db,
    provider: &str,
    cred: &Credential,
    client: &reqwest::Client,
) -> Result<Option<Credential>> {
    let now = chrono::Utc::now().timestamp_millis();
    if !cred.needs_refresh(now) {
        return Ok(None);
    }
    let Credential::Oauth(oauth) = cred else { return Ok(None) };
    let Some(refresh_token) = oauth.refresh.as_deref() else { return Ok(None) };
    let refreshed = match provider {
        "anthropic" => crate::auth::oauth::anthropic::refresh(client, refresh_token).await?,
        "google-antigravity" | "google-gemini-cli" => {
            crate::auth::oauth::google::refresh(client, refresh_token, oauth).await?
        }
        _ => return Ok(None),
    };
    let new_cred = Credential::Oauth(refreshed);
    cred_store::upsert(db, provider, new_cred.clone())?;
    Ok(Some(new_cred))
}

File: src/usage/orchestrator.ts
/**
 * Orchestrates live usage fetches: walks atop's stored credentials for each
 * supported provider, refreshes expired OAuth tokens, and returns the union of
 * UsageReports plus a per-credential error log (so the UI can show *why* a
 * provider is empty instead of looking blank). Port of `orchestrator.rs`.
 */
import type { Database } from "bun:sqlite";
import type { Credential } from "../auth/types.ts";
import type { UsageReport } from "./types.ts";
import { accountLabel, needsRefresh } from "../auth/types.ts";
import { allProviders, load, upsert } from "../auth/store.ts";
import { fetcherFor } from "./fetcher.ts";
import { refresh as anthropicRefresh } from "../auth/oauth/anthropic.ts";
import { refresh as googleRefresh } from "../auth/oauth/google.ts";
import { refresh as openaiCodexRefresh } from "../auth/oauth/openaiCodex.ts";
import { refresh as minimaxRefresh } from "../auth/oauth/minimax.ts";
import { refresh as kimiRefresh } from "../auth/oauth/kimi.ts";
import { refresh as xaiRefresh } from "../auth/oauth/xai.ts";
import { refresh as opencodeRefresh } from "../auth/oauth/opencode.ts";

export interface FetchError {
  provider: string;
  account: string;
  message: string;
}

export interface FetchResult {
  reports: UsageReport[];
  errors: FetchError[];
}

function refreshFor(provider: string): ((cred: Credential) => Promise<Credential>) | undefined {
  switch (provider) {
    case "anthropic":
      return anthropicRefresh;
    case "google-antigravity":
    case "google-gemini-cli":
      return googleRefresh;
    case "openai-codex":
      return openaiCodexRefresh;
    case "minimax-code":
    case "minimax-code-cn":
      return minimaxRefresh;
    case "kimi-code":
      return kimiRefresh;
    case "xai-oauth":
      return xaiRefresh;
    case "opencode-go":
      return opencodeRefresh;
    default:
      return undefined;
  }
}

const errMessage = (e: unknown): string => (e instanceof Error ? e.message : String(e));

/**
 * Fetch live usage for every credential atop has stored for the given
 * providers (default: all providers with stored credentials). A single failed
 * credential never blocks the others.
 */
export async function fetchAll(
  db: Database,
  dataDir: string,
  opts?: { provider?: string },
): Promise<FetchResult> {
  const providers = opts?.provider ? [opts.provider] : allProviders(dataDir);
  const errors: FetchError[] = [];
  const pending: Promise<UsageReport | null>[] = [];
  const now = Date.now();

  for (const provider of providers) {
    for (const original of load(dataDir, provider)) {
      let cred = original;
      if (needsRefresh(cred, now) && cred.type === "oauth" && cred.refresh) {
        const doRefresh = refreshFor(provider);
        if (doRefresh) {
          try {
            cred = await doRefresh(cred);
            upsert(dataDir, provider, cred);
          } catch (e) {
            errors.push({ provider, account: accountLabel(original), message: errMessage(e) });
            continue;
          }
        }
      }
      const account = accountLabel(cred);
      const fetcher = fetcherFor(provider);
      if (!fetcher) {
        errors.push({ provider, account, message: "no fetcher registered for this provider" });
        continue;
      }
      pending.push(
        fetcher.fetch(cred, db).then((report) => {
          if (report === null) throw new Error("provider returned no usage data");
          return report;
        }).catch((e) => {
          errors.push({ provider, account, message: errMessage(e) });
          return null;
        }),
      );
    }
  }

  const settled = await Promise.all(pending);
  const reports = settled.filter((r): r is UsageReport => r !== null);
  return { reports, errors };
}

/** Per-provider freshest report (used by the TUI). */
export function freshest(reports: UsageReport[]): Map<string, UsageReport> {
  const by = new Map<string, UsageReport>();
  for (const r of reports) {
    const prev = by.get(r.provider);
    if (!prev || prev.fetchedAt < r.fetchedAt) by.set(r.provider, r);
  }
  return by;
}
